### DISCUSSION TITLE
Need help using rastervision with pre-classified .tiff files

### DISCUSSION THREAD

**Discussion by andrebiehl** [azavea/raster-vision#1785] [General]
Hello! I'm working with a geospatial .tiff dataset where each pixel is a pre-classified value between 0-4 (5 values). For a proof of concept, I tried to train rastervision on one single .tiff file, which I split 80/20 pixel-wise into training and validation. I'm getting an error when running this cell: 

# Define scene configurations
train_scene = SceneConfig(
    id='train_scene',
    raster_source=RasterioSourceConfig(uris=['train.tif']),
    label_source=RasterioSourceConfig(uris=['train.tif']),
)

val_scene = SceneConfig(
    id='val_scene',
    raster_source=RasterioSourceConfig(uris=['val.tif']),
    label_source=RasterioSourceConfig(uris=['val.tif']),
)

Note: train.tif is includes the pixels for training and val.tif includes the pixels for validation. 

The error is because I'm trying to use the RasterioSourceConfig for both the raster source and label source, which is not allowed because they are expected to be different types of data (features vs labels).

If my labels in the same geospatial TIFF file as my input data (i.e., different bands in the TIFF file represent different labels), I will probably need to use a different approach for creating my LabelSourceConfig.

Any advice?
[rastervisionTest.pdf](https://github.com/azavea/raster-vision/files/11626578/rastervisionTest.pdf)

**Comment by AdeelH** [comment]
Let's assume `train.tif` has 4 bands with the first 3 being the image and the last one being labels. Then you can do the following:
```py
raster_source_img = RasterioSourceConfig(uris=['train.tif'], channel_order=[0, 1, 2])
raster_source_label = RasterioSourceConfig(uris=['train.tif'], channel_order=[3])
train_scene = SceneConfig(
    id='train_scene',
    raster_source=raster_source_img,
    label_source=SemanticSegmentationLabelSourceConfig(raster_source=raster_source_label),
)
```

**Reply by andrebiehl** [reply]
Thank you, but that's not exactly what I meant. My data is already labeled: each pixel contains one value from 0 to 4 (5 possible categories). So the pixel doesn't contain any bands, just labels. I'm trying to find a way to turn these labels into a vector representation.

**Reply by AdeelH** [reply]
So you just want to vectorize these raster labels? Have these labels been produced through Raster Vision?

**Reply by andrebiehl** [reply]
That's correct. They have not been produced by raster vision, they're pre-classified from another algorithm into 5 categories. I'm having trouble finding a way to vectorize these large tiff files and I thought raster vision would be helpful. To be clear, I don't need raster vision to classify the images, I just want to use its vectorization functionality. I already tried using numpy and related frameworks but it's too slow as it runs on CPU and not GPU

**Reply by AdeelH** [reply]
I see. Unfortunately, Raster Vision does not make use of the GPU for this either. It just uses [`rasterio.features.shapes()`](https://rasterio.readthedocs.io/en/stable/api/rasterio.features.html#rasterio.features.shapes) to do the vectorization.

One reason why vectorization can take very long is if the predictions appear as a large number of very small and scattered regions (i.e. it would take a huge number of polygons to capture them). RV allows you to do some filtering to get rid of regions below a certain radius using the [`denoise()`](https://docs.rastervision.io/en/0.20/api_reference/_generated/rastervision.core.data.utils.vectorization.denoise.html#denoise) funtion (which is just a [simple wrapper around an OpenCV-based morphological transform](https://docs.rastervision.io/en/0.20/_modules/rastervision/core/data/utils/vectorization.html#denoise)).


If you want to use RV for this, you can try the following:
- Rename your TIFF to `labels.tif`
- Assuming RV v0.20 and `labels.tif` is in dir `/some/dir/`, something like the following _should_ work:
```py
from rastervision.core.data import (
    ClassConfig, SemanticSegmentationLabelStore, RasterioCRSTransformer,
    PolygonVectorOutputConfig)

class_config = ClassConfig(
    names=['class1', 'class2', 'class3', 'class4', 'class5'])
crs_transformer = RasterioCRSTransformer.from_uri('/some/dir/labels.tif')
vector_outputs = [
    PolygonVectorOutputConfig(
        class_id=i,
        uri=f'/some/dir/vector_outputs/{class_config.get_name(i)}.json',
        denoise=8) for i in range(len(class_config))
]

label_store = SemanticSegmentationLabelStore(
    uri='/some/dir/',
    crs_transformer=crs_transformer,
    class_config=class_config,
    vector_outputs=vector_outputs,
    discrete_output=False,
    smooth_output=False)
labels = label_store.get_labels()
label_store.write_vector_outputs(labels)
```

If instead you just want to emulate how RV does it, you could try to adapt the following piece of code:
https://github.com/azavea/raster-vision/blob/master/rastervision_core/rastervision/core/data/label_store/semantic_segmentation_label_store.py#L297-L315

**Reply by andrebiehl** [reply]
Thank you so much, denoise would indeed be helpful for our use case. I appreciate your help.
